float64[2] desired_input
int64 mode
